### PR TITLE
 Fix Issue 19645 - Default parameters not checked for @safe

### PR DESCRIPTION
**Pull request by RazvanN7**
Default parameters are not checked for `@safe`/`pure`/`@nogc`/`nothrow`. This is mainly because dmd implemented these checks only for expressions that are in a functions' scope. Before this patch dmd created a normal scope (not function scope) for parameters to be evaluated; this led to default parameters not being checked for `@safe`/`pure`/`@nogc`/`nothrow`.

To solve the issue I created a mock-up function declaration to trick the compiler that the parameters are in a functions's scope. It would have been ideal to use the initial function declaration for this, but there is no way of obtaining it from a TypeFunction. 

`nothrow` is not handled in this PR as it is a bit different.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

File: src/dmd/id.d [hunk 1]
```diff
@@ -314,6 +314,7 @@ immutable Msgtable[] msgtable =
     { "dup" },
     { "_aaApply" },
     { "_aaApply2" },
+    { "__tmpFunc"},
```
Comment (thewilsonator) [inline src/dmd/id.d]: Would be nice to give a bit of a description here.

Comment (RazvanN7) [reply src/dmd/id.d]: temporary function used as mock-up scope to evaluate function parameters ;)

Comment (thewilsonator) [reply src/dmd/id.d]: I know that _now_, I might not in 6 months time.

**Comment by ZombineDev** [comment]
Shouldn't the function attributes of the default argument expressions be judged based on the caller and not the function their declared in?

**Comment by jacob-carlborg** [comment]
> It would have been ideal to use the initial function declaration for this, but there is no way of obtaining it from a TypeFunction.

Why don't you add that?

File: src/dmd/typesem.d [hunk 2]
```diff
@@ -1206,6 +1206,10 @@ extern(C++) Type typeSemantic(Type t, Loc loc, Scope* sc)
             argsc.stc = 0; // don't inherit storage class
             argsc.protection = Prot(Prot.Kind.public_);
             argsc.func = null;
+            Loc tloc = Loc();
+            auto mockFunc = new FuncDeclaration(tloc, tloc, Id.__tmpFunc, sc.stc, new TypeFunction(ParameterList(), Type.tvoid, LINK.d, sc.stc));
```
Comment (jacob-carlborg) [inline src/dmd/typesem.d]: Giving the function declaration and identifier will add it to the symbol table which might make it accessible to user code.

**Pull request closed without merging**